File: RepoTransBench/target_projects/C/Python/mbuesch_razer/tests/original/test_deathadder2013.py
import pytest

# Mocks
class RazerMouseUsbCtx:
    def __init__(self):
        self.h = 1

class RazerMouse:
    def __init__(self):
        self.usb_ctx = None

class Deathadder2013Private:
    def __init__(self, m):
        self.m = m

class Deathadder2013Command:
    def __init__(self):
        # There might be a specific size for the structure in C. We'll simulate 16 bytes.
        self.buf = bytearray(16)
        self.command = 0
        self.request = 0

def razer_error(msg, *args):
    pass

# Simulate control transfer behavior
class ControlTransferMock:
    def __init__(self):
        self.call = 0
    def libusb_control_transfer(self, h, flags, request, command, zero, buf, size, timeout):
        self.call += 1
        if self.call == 1:
            return -1
        return size

def deathadder2013_command_init(cmd):
    cmd.buf = bytearray([0] * len(cmd.buf))
    # In C, all members set to zero (already done above)
    cmd.command = 0
    cmd.request = 0

def deathadder2013_usb_write(priv, a, b, buf, size, control_mock=None):
    if control_mock is None:
        control_mock = ControlTransferMock()
    # First call always fails
    rc = control_mock.libusb_control_transfer(
        priv.m.usb_ctx.h, 0, 0, 0, 0, buf.buf, size, 500
    )
    if rc < 0:
        return -1
    return 0

def test_command_init():
    cmd = Deathadder2013Command()
    # fill with 0xFF
    cmd.buf = bytearray([0xFF] * len(cmd.buf))
    deathadder2013_command_init(cmd)
    assert all(b == 0 for b in cmd.buf)

def test_usb_write_error_and_success():
    ctx = RazerMouseUsbCtx()
    m = RazerMouse()
    m.usb_ctx = ctx
    priv = Deathadder2013Private(m)
    buf = Deathadder2013Command()
    # Mock control transfer so first call fails
    control_mock = ControlTransferMock()
    rc = deathadder2013_usb_write(priv, 1, 2, buf, len(buf.buf), control_mock)
    assert rc == -1
    # In the C test, next call would succeed, but we only test the error-case

File: RepoTransBench/target_projects/C/Rust/mbuesch_razer/tests/original/test_deathadder2013.rs
/*
 Basic tests for DeathAdder2013 command _init and usb_write/read error handling
 Mock USB/razer_error
*/
use std::cell::Cell;
use std::sync::Mutex;

use std::sync::atomic::{AtomicUsize, Ordering};

#[repr(C)]
struct Deathadder2013Command {
    data: [u8; 8],
    command: u8,
    request: u8,
}
impl Default for Deathadder2013Command {
    fn default() -> Self {
        Deathadder2013Command {
            data: [0; 8],
            command: 0,
            request: 0,
        }
    }
}

struct RazerMouseUsbCtx {
    h: i32,
}

struct RazerMouse {
    usb_ctx: *mut RazerMouseUsbCtx,
}
struct Deathadder2013Private<'a> {
    m: &'a mut RazerMouse,
}

static CALL_COUNT: AtomicUsize = AtomicUsize::new(0);

fn razer_error(_msg: &str) {
    // ignore
}

fn libusb_control_transfer(
    _h: i32,
    _flags: i32,
    _request: i32,
    _command: i32,
    _zero: i32,
    _buf: &mut [u8],
    size: usize,
    _timeout: i32,
) -> i32 {
    // Simulate usb fail on first attempt, succeed on second for read case
    let call = CALL_COUNT.fetch_add(1, Ordering::SeqCst);
    if call == 0 {
        return -1;
    }
    size as i32
}
const RAZER_USB_TIMEOUT: i32 = 500;

fn deathadder2013_command_init(cmd: &mut Deathadder2013Command) {
    let size = std::mem::size_of_val(cmd);
    let ptr = cmd as *mut _ as *mut u8;
    unsafe { std::ptr::write_bytes(ptr, 0, size); }
}

fn deathadder2013_usb_write<'a>(
    _priv: &Deathadder2013Private<'a>,
    _r1: i32,
    _r2: i32,
    _buf: &Deathadder2013Command,
    _size: usize,
) -> i32 {
    // Always fail, as in C test
    -1
}

#[test]
fn test_command_init() {
    let mut cmd = Deathadder2013Command {
        data: [0xFF; 8],
        command: 0xFF,
        request: 0xFF,
    };
    deathadder2013_command_init(&mut cmd);
    let ptr = &cmd as *const _ as *const u8;
    let size = std::mem::size_of_val(&cmd);
    for i in 0..size {
        unsafe {
            assert_eq!(*ptr.add(i), 0);
        }
    }
}

#[test]
fn test_usb_write_error_and_success() {
    let mut ctx = RazerMouseUsbCtx { h: 1 };
    let mut m = RazerMouse { usb_ctx: &mut ctx as *mut _ };
    let priv_struct = Deathadder2013Private { m: &mut m };
    let buf = Deathadder2013Command::default();
    let rc = deathadder2013_usb_write(&priv_struct, 1, 2, &buf, std::mem::size_of_val(&buf));
    assert_eq!(rc, -1); // always fails, since our mock returns -1 first call
    // next call would succeed, but we continue error-case
}